app/src/main/java/com/sdp13epfl2021/projmag/database/UserProfileDatabase.kt

```
@@ -0,0 +1,142 @@
+                            else -> Gender.OTHER
+                        }
+                        val phoneNumber = document["phoneNumber"] as? String
+                        val roleString = document["role"] as? String
+                        val sciperLong = document["sciper"]
+
+                        val role = when (roleString) {
+                            Role.TEACHER.name -> Role.TEACHER
+                            Role.STUDENT.name -> Role.STUDENT
+                            else -> Role.OTHER
+                        }
+
+
+                        var age: Int = try {
+                            Integer.valueOf(ageLong.toString())
+                        } catch (excep: NumberFormatException) {
+                            0
+                        }
+
+                        var sciper: Int = try {
+                            Integer.valueOf(sciperLong.toString())
+                        } catch (exep: NumberFormatException) {
+                            0
+                        }
+
+
+                        when (val resProfile = ImmutableProfile.build(
+                            lastName ?: "null",
+                            firstName ?: "null",
+                            age,
+                            gender,
+                            sciper,
+                            phoneNumber ?: "null",
+                            role
+                        )) {
+                            is Success -> {
+                                onSuccess(resProfile.value)
+                            }
+                            is Failure -> {
+                                Log.d(TAG, "Failure reason : ${resProfile.reason}")
+                            }
+                        }
+
+                    } else {
+                        Log.d(TAG, "No such document")
+                    }
+                }
+                .addOnFailureListener { exception ->
+                    Log.d(TAG, "get failed with ", exception)
```

Instead of just using logging, adding a new argument with a failure callback could be great. (Such as in other database function)